tls/s2n_kem.c

```diff
@@ -0,0 +1,113 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "stuffer/s2n_stuffer.h"
+
+#include "tls/s2n_kem.h"
+
+#include "utils/s2n_safety.h"
+#include "utils/s2n_mem.h"
+
+int s2n_kem_generate_keypair(struct s2n_kem_keypair *kem_keys)
+{
+    const struct s2n_kem *kem = kem_keys->negotiated_kem;
+    notnull_check(kem->generate_keypair);
+
+    eq_check(kem_keys->public_key.size, kem->public_key_length);
+    notnull_check(kem_keys->public_key.data);
+
+    // The private key is needed for client_key_recv and must be saved
+    GUARD(s2n_alloc(&kem_keys->private_key, kem->private_key_length));
+
+    GUARD(kem->generate_keypair(kem_keys->public_key.data, kem_keys->private_key.data));
+    return 0;
+}
+
+int s2n_kem_encapsulate(const struct s2n_kem_keypair *kem_keys, struct s2n_blob *shared_secret,
+                        struct s2n_blob *ciphertext)
+{
+    const struct s2n_kem *kem = kem_keys->negotiated_kem;
+    notnull_check(kem->encapsulate);
+
+    eq_check(kem_keys->public_key.size, kem->public_key_length);
+    notnull_check(kem_keys->public_key.data);
+
+    eq_check(ciphertext->size, kem->ciphertext_length);
+    notnull_check(ciphertext->data);
+
+    GUARD(s2n_alloc(shared_secret, kem->shared_secret_key_length));
+
+    GUARD(kem->encapsulate(ciphertext->data, shared_secret->data, kem_keys->public_key.data));
+    return 0;
+}
+
+int s2n_kem_decapsulate(const struct s2n_kem_keypair *kem_keys, struct s2n_blob *shared_secret,
+                        const struct s2n_blob *ciphertext)
+{
+    const struct s2n_kem *kem = kem_keys->negotiated_kem;
+    notnull_check(kem->decapsulate);
+
+    eq_check(kem_keys->private_key.size, kem->private_key_length);
+    notnull_check(kem_keys->private_key.data);
+
+    eq_check(ciphertext->size, kem->ciphertext_length);
+    notnull_check(ciphertext->data);
+
+    GUARD(s2n_alloc(shared_secret, kem_keys->negotiated_kem->shared_secret_key_length));
+
+    GUARD(kem->decapsulate(shared_secret->data, ciphertext->data, kem_keys->private_key.data));
+    return 0;
+}
+
+int s2n_kem_find_supported_kem(struct s2n_blob *client_kem_ids, const struct s2n_kem *supported_kems,
+                               const int num_supported_kems, const struct s2n_kem **matching_kem)
+{
+    struct s2n_stuffer kem_name_in = {{0}};
+
+    GUARD(s2n_stuffer_init(&kem_name_in, client_kem_ids));
+    GUARD(s2n_stuffer_write(&kem_name_in, client_kem_ids));
+
+    for (int i = 0; i < num_supported_kems; i++) {
+        const struct s2n_kem candidate_server_kem_name = supported_kems[i];
+        for (int j = 0; j < client_kem_ids->size; j++) {
+            kem_extension_size kem_id;
+            GUARD(s2n_stuffer_read_uint8(&kem_name_in, &kem_id));
+
+            if (candidate_server_kem_name.kem_extension_id == kem_id) {
+                *matching_kem = &supported_kems[i];
+                return 0;
+            }
+        }
```

Comment:
if client_key_ids is just a simple array of bytes, I'm not sure there's much value in using a stuffer here. This might be a rare case where it's better to go without.  Is it just a simple array?

Reply:
Yes, the client sends a list of bytes that indicate what KEM parameters they support. For BIKE the extension is:
```
enum {
     BIKE1-Level1 (1), BIKE1-Level3 (2), BIKE1-Level5 (3), BIKE2-Level1 (4), BIKE2-Level3 (5), BIKE2-Level5 (6), BIKE3-Level1 (7), BIKE3-Level3 (8), BIKE3-Level5 (9)
} NamedBIKEKEM (2^8-1);
struct {
    NamedBIKEKEM bike_parameter_list<1..2^8-1>
} BIKEParameterList;
```
A client might send an extension like:
```
FE 01 03 02 01 04
```
`FE 01` is BIKE supported param extension, `03` is the total length of the extension, `02` is the length of the BIKEParameterList, `01` and `04` is BIKE1-Level1 and BIKE2-Level1. This method is designed to be called with just the list of supported parameters `{0x01, 0x04}`.

For the full specification and to compare what SIKE looks like check out the description in https://github.com/awslabs/s2n/issues/904.

I will look into how this looks using a simple array.